rtsp_to_hls/hls_proxy_api/lib/connection/connection_manager.ex

```diff
@@ -0,0 +1,399 @@
+           {:ok, connection_status} <- start_pipeline(connection_status),
+           {:ok, connection_status} <- start_keep_alive(connection_status),
+           :ok <- play(connection_status) do
+        connection_status = %{connection_status | status: :ok}
+
+        Logger.warn(~s"""
+        ConnectionManager_#{connection_status.stream.id} processes:
+          RTSP session: #{inspect(connection_status.rtsp_session)},
+          Membrane Pipeline: #{inspect(connection_status.pipeline)},
+          RTSP keep alive: #{inspect(connection_status.keep_alive)}
+        """)
+
+        {:ok, connection_status}
+      else
+        {:error, error_message} ->
+          Logger.warn("ConnectionManager_#{id}: Connection failed: #{inspect(error_message)}")
+          {:backoff, @delay, connection_status}
+      end
+    end
+  end
+
+  @impl true
+  def disconnect(
+        message,
+        %ConnectionStatus{
+          stream: stream,
+          pipeline_options: pipeline_options
+        } = connection_status
+      ) do
+    %Stream{id: id} = stream
+    Logger.debug("ConnectionManager_#{id}: Disconnecting: #{message}")
+
+    kill_children(connection_status)
+
+    File.rm_rf(pipeline_options[:output_path])
+
+    connection_status = %{
+      connection_status
+      | status: :error,
+        pipeline: nil,
+        rtsp_session: nil,
+        keep_alive: nil
+    }
+
+    case message do
+      :close ->
+        PortAgent.remove(pipeline_options[:port])
+        {:stop, :shutdown, connection_status}
+
+      :reload ->
+        {:connect, :reload, connection_status}
+
+      {:error, error_message} ->
+        Logger.error("ConnectionManager_#{id}: Error: #{inspect(error_message)}")
+        {:backoff, @delay, connection_status}
+    end
+  end
+
+  defp kill_children(%ConnectionStatus{
+          pipeline: pipeline,
+          keep_alive: keep_alive,
+          rtsp_session: rtsp_session
+        }) do
+    if !is_nil(pipeline) and Process.alive?(pipeline),
+      do: Membrane.Pipeline.terminate(pipeline)
+
+    if !is_nil(keep_alive) and Process.alive?(keep_alive), do: GenServer.stop(keep_alive, :normal)
+
+    if !is_nil(rtsp_session) and Process.alive?(rtsp_session), do: RTSP.close(rtsp_session)
+  end
+
+  @impl true
+  def handle_call(:close, _from, connection_status) do
+    {:disconnect, :close, connection_status}
+  end
+
+  @impl true
+  def handle_call({:reload, new_connection_status}, _from, connection_status) do
+    {:disconnect, :reload, Map.merge(connection_status, new_connection_status)}
+  end
+
+  @impl true
+  def handle_call(:status, _from, %{status: status} = connection_status) do
+    {:reply, status, connection_status}
+  end
+
+  @impl true
+  def handle_info(
+        {:DOWN, _ref, :process, pid, reason},
+        %ConnectionStatus{
+          rtsp_session: rtsp_session,
+          pipeline: pipeline,
+          keep_alive: keep_alive
+        } = connection_status
+      )
+      when reason != :normal do
```

Comment:
how about removing that guard: `when reason != :normal`, and putting this function case matching below the one with `when reason == normal`?